### PR TITLE
Rework Processes assessor to use Cpanel module

### PR DESCRIPTION
**Pull request by ColMelvin**
Case PIG-2858: To encourage re-use, the logic to check for outdated
processes was copied to cPanel & WHM.  It was also improved to further
encourage adoption.  Modify this assessor to use the new module to gain
these improvements.

### CODE REVIEW

File: pkg/Cpanel/Security/Advisor/Assessors/Processes.pm [hunk 1]
```diff
@@ -46,61 +46,145 @@ sub generate_advice {
...
+                    'Install the ‘[_1]’ command to get notifications when executables are updated but the existing processes are not restarted.',
+                    $err->get('service'),
+                ),
+            );
+            return;    # Cannot check any other cases, so abort.
+        }
+        elsif ( !ref $err || !$err->isa('Cpanel::Exception::Unsupported') ) {
+            $self->add_warn_advice(
+                key  => 'Processes_error_while_checking_reboot',
+                text => $self->_lh->maketext( 'An error occurred while attempting to check whether to reboot: [_1]', Cpanel::Exception::get_string($err) ),
```
Comment (ryeddula) [inline pkg/Cpanel/Security/Advisor/Assessors/Processes.pm]: Would recommend using 'get_string_no_id' instead as get_string give a random "error id."

File: pkg/Cpanel/Security/Advisor/Assessors/Processes.pm [hunk 2]
```diff
@@ -46,61 +46,145 @@ sub generate_advice {
...
-                'key'      => 'Processes_detected_running_from_outdated_executables',
-                text       => $self->_lh->maketext('Detected processes that are running outdated binary executables.'),
-                suggestion => $self->_lh->maketext(
-                    'Reboot the system in the “[output,url,_1,Graceful Server Reboot,_2,_3]” area.  Alternatively, [asis,SSH] into this server and run ‘[_4]’, then manually restart each of the listed processes.',
+    my @services = eval { Cpanel::ProcessCheck::Outdated::outdated_services() };
+    if ( my $err = $@ ) {
+        if ( !ref $err || !$err->isa('Cpanel::Exception::Unsupported') ) {
+            $self->add_warn_advice(
+                key  => 'Processes_error_while_checking_running_services',
+                text => $self->_lh->maketext( 'An error occurred while attempting to check whether running services are up-to-date: [_1]', Cpanel::Exception::get_string($err) ),
```
Comment (ryeddula) [inline pkg/Cpanel/Security/Advisor/Assessors/Processes.pm]: Same no_id

File: pkg/Cpanel/Security/Advisor/Assessors/Processes.pm [hunk 3]
```diff
@@ -46,61 +46,145 @@ sub generate_advice {
...
                     'target',
                     '_blank',
-                    $exec,
                 ),
-            );
-        }
-        elsif ( $proc->CHILD_ERROR() ) {
+            ),
+        );
+        return;    # No need to check further.
```
Comment (ryeddula) [inline pkg/Cpanel/Security/Advisor/Assessors/Processes.pm]: Hmm, why stop checking the services come back as needing a restart? Will the outdated_processes() checks return the same set or something?

File: pkg/Cpanel/Security/Advisor/Assessors/Processes.pm (hunk 1 above)
Comment (ColMelvin) [inline pkg/Cpanel/Security/Advisor/Assessors/Processes.pm]: Good idea

File: pkg/Cpanel/Security/Advisor/Assessors/Processes.pm [hunk 4]
```diff
@@ -46,61 +46,145 @@ sub generate_advice {
...
+    $self->add_good_advice(
+        key  => 'Processes_none_with_outdated_executables',
+        text => $self->_lh->maketext('No processes with outdated binaries detected.')
+    );
+
     return 1;
 }
 
+# Do this to work around bad perltidy concatenation rules.
+sub _make_unordered_list {
```
Comment (royerboat) [inline pkg/Cpanel/Security/Advisor/Assessors/Processes.pm]: Not a fan of writing HTML directly in the perl. Not sure how you would avoid it in this scenario though.

File: pkg/Cpanel/Security/Advisor/Assessors/Processes.pm (hunk 3 above)
Comment (ColMelvin) [inline pkg/Cpanel/Security/Advisor/Assessors/Processes.pm]: Many of the PIDs listed by `outdated_processes()` map back to a service listed in `outdated_services()`.  However, not all of them will necessarily map back.  We would need a way to filter out the PIDs that are not associated with a service (which is probably a thing).  If we could reliably filter these, I'd be down for displaying both messages at once.

If we didn't filter out service PIDs, we might get into an instance like the following: Exim gets updated, so we'll tell them to restart exim.service and PID 1234 - exim.  By taking action on the first, you should not take action on the second.

File: pkg/Cpanel/Security/Advisor/Assessors/Processes.pm (hunk 4 above)
Comment (ColMelvin) [inline pkg/Cpanel/Security/Advisor/Assessors/Processes.pm]: Yeah, we'd need to revisit `add_bad_advice` and friends, which would be an improvement.

**Pull request merged**